### PR TITLE
RetryPolicy enhancements

### LINKED ISSUE #59: Allow to dynamically define a retry delay based on the result of the Promise passed to execute
#### Describe the feature

A common use case for a circuit breaker is to retry or circuit break remote HTTP calls done by a web/HTTP client.
HTTP servers use the `Retry-After` header to dynamically suggest to clients to retry after a defined period of time during their unavailability.
The current circuit breaker API allows defining the retry period at creation time only, so it's not possible to define it dynamically based on whatever the downstream service suggests.

#### Use cases

(covered above)

#### Contribution

Who should implement this feature ? are you volunteering for implementing this feature or
do you know that is able and willing implement this feature ?

I can help but I'm not entirely sure about a Vert.x idiomatic API for exposing this.

### LINKED ISSUE #60: Provide common retry policies like constant, linear and exponential
#### Describe the feature

Retry policies can be configured using the `retryPolicy` method.
it's common to use well-known retry policies like constant, linear, or exponential backoff delays for this policy.

#### Use cases

- Use an already provided constant backoff delay given a delay in milliseconds as retry policy function
- Use an already provided linear backoff delay given a delay in milliseconds as retry policy function
- Use an already provided exponential backoff delay given a delay in milliseconds as retry policy function

#### Contribution

Who should implement this feature ? are you volunteering for implementing this feature or
do you know that is able and willing implement this feature ?

Yes, I can implement this feature.

### PR DESCRIPTION
**Pull request by tsegismont**
Closes #59, closes #60

### CODE REVIEW

**Comment by tsegismont** [comment]
@jponge @pierDipi would you mind taking a look to the proposed API?

The PR is not finished yet, when I get feedback about the API I will add tests and update the documentation.

File: src/test/java/io/vertx/circuitbreaker/impl/RetryPolicyTest.java [hunk 1]
```diff
@@ -31,31 +31,31 @@ public void tearDown() {
 
   @Test
   public void testWithRetryPolicy() {
-    runRetryPolicyTest(retry -> retry * 100L);
+    runRetryPolicyTest(RetryPolicy.linearDelay(100, 10000));
```
Comment (jponge) [inline src/test/java/io/vertx/circuitbreaker/impl/RetryPolicyTest.java:34]: You might also want to test that exceptions are thrown on incorrect parameters.

**Review by jponge — COMMENTED** [review]
Factoring out of the policy from the circuit-breaker interface is a good idea 👍

**Comment by tsegismont** [comment]
@jponge @pierDipi I've added a test which shows how to configure retry policy when working with an HTTP server that returns `Retry-After` header.

**Pull request merged**